### PR TITLE
Update DeviceCodeCredential parameters

### LINKED ISSUE #11877: Add "Use Console" Option to DeviceCodeCredential
Investigate adding a "Use Console" option for Device Code Credential

### PR DESCRIPTION
**Pull request by catalinaperalta**
This PR updates DeviceCodeCredential to have only optional parameters. 

Fixes #11877

### CODE REVIEW

File: sdk/azidentity/device_code_credential.go [hunk 1]
```diff
@@ -15,27 +16,47 @@ const (
 	deviceCodeGrantType = "urn:ietf:params:oauth:grant-type:device_code"
 )
 
+type DeviceCodeCredentialOptions struct {
+	TenantID   *string                 // Gets the Azure Active Directory tenant (directory) ID of the service principal
+	ClientID   *string                 // Gets the client (application) ID of the service principal
+	UserPrompt *func(string)           // The callback function used to send the login message back to the user
```
Comment (jhendrixMSFT) [inline sdk/azidentity/device_code_credential.go:22]: Funcs are implicitly `nil`able so no need for `*`.

File: sdk/azidentity/device_code_credential.go [hunk 2]
```diff
@@ -15,27 +16,47 @@ const (
...
-// clientID: The client (application) ID of the service principal.
-// callback: The callback function used to send the login message back to the user
-// options: Options used to configure the management of the requests sent to Azure Active Directory.
-func NewDeviceCodeCredential(tenantID string, clientID string, callback func(string), options *TokenCredentialOptions) (*DeviceCodeCredential, error) {
-	c, err := newAADIdentityClient(options)
+// options: Options used to configure the management of the requests sent to Azure Active Directory, please see DeviceCodeCredentialOptions for a description of each field.
+func NewDeviceCodeCredential(options *DeviceCodeCredentialOptions) (*DeviceCodeCredential, error) {
+	tenantID := "organizations"
+	clientID := developerSignOnClientID
+	userPrompt := handler
```
Comment (jhendrixMSFT) [inline sdk/azidentity/device_code_credential.go:41]: You can simply do this.
```go
userPrompt := func(s string) {
	fmt.Println(s)
}
```
And omit the definition of `handler()` below.

File: sdk/azidentity/device_code_credential.go [hunk 3]
```diff
@@ -15,27 +16,47 @@ const (
...
 // NewDeviceCodeCredential constructs a new DeviceCodeCredential used to authenticate against Azure Active Directory with a device code.
-// tenantID: The Azure Active Directory tenant (directory) ID of the service principal. If none is set then the default value ("organizations") will be used in place of the tenantID.
-// clientID: The client (application) ID of the service principal.
-// callback: The callback function used to send the login message back to the user
-// options: Options used to configure the management of the requests sent to Azure Active Directory.
-func NewDeviceCodeCredential(tenantID string, clientID string, callback func(string), options *TokenCredentialOptions) (*DeviceCodeCredential, error) {
-	c, err := newAADIdentityClient(options)
+// options: Options used to configure the management of the requests sent to Azure Active Directory, please see DeviceCodeCredentialOptions for a description of each field.
+func NewDeviceCodeCredential(options *DeviceCodeCredentialOptions) (*DeviceCodeCredential, error) {
+	tenantID := "organizations"
```
Comment (jhendrixMSFT) [inline sdk/azidentity/device_code_credential.go:39]: All these defaults should be moved to a `DefaultDeviceCodeCredentialOptions()` func.

File: sdk/azidentity/device_code_credential.go [hunk 4]
```diff
@@ -15,27 +16,47 @@ const (
 	deviceCodeGrantType = "urn:ietf:params:oauth:grant-type:device_code"
 )
 
+type DeviceCodeCredentialOptions struct {
+	TenantID   *string                 // Gets the Azure Active Directory tenant (directory) ID of the service principal
```
Comment (jhendrixMSFT) [inline sdk/azidentity/device_code_credential.go:20]: I believe `TenantID` and `ClientID` won't be pointer-to-string.

Comment (catalinaperalta) [reply sdk/azidentity/device_code_credential.go:20]: Done

File: sdk/azidentity/device_code_credential.go (hunk 3 above)
Comment (catalinaperalta) [reply sdk/azidentity/device_code_credential.go]: Done

File: sdk/azidentity/device_code_credential.go (hunk 2 above)
Comment (catalinaperalta) [reply sdk/azidentity/device_code_credential.go]: Done

File: sdk/azidentity/device_code_credential.go (hunk 1 above)
Comment (catalinaperalta) [reply sdk/azidentity/device_code_credential.go]: Done

File: sdk/azidentity/device_code_credential.go [hunk 5]
```diff
@@ -15,27 +16,47 @@ const (
 	deviceCodeGrantType = "urn:ietf:params:oauth:grant-type:device_code"
 )
 
+// DeviceCodeCredentialOptions provide options that can configure DeviceCodeCredential instead of using the default values.
+type DeviceCodeCredentialOptions struct {
+	TenantID   string                  // Gets the Azure Active Directory tenant (directory) ID of the service principal
```
Comment (jhendrixMSFT) [inline sdk/azidentity/device_code_credential.go:21]: Doc comments go above field declarations.

File: sdk/azidentity/device_code_credential.go [hunk 6]
```diff
@@ -15,27 +16,47 @@ const (
...
 	clientID     string       // Gets the client (application) ID of the service principal
-	callback     func(string) // Sends the user a message with a verification URL and device code to sign in to the login server
+	userPrompt   func(string) // Sends the user a message with a verification URL and device code to sign in to the login server
 	refreshToken string       // Gets the refresh token sent from the service and will be used to retreive new access tokens after the initial request for a token. Thread safety for updates is handled in the AuthenticationPolicy since only one goroutine will be updating at a time
 }
 
+// DefaultDeviceCodeCredentialOptions provides the default settings for DeviceCodeCredential.
+func DefaultDeviceCodeCredentialOptions() DeviceCodeCredentialOptions {
+	return DeviceCodeCredentialOptions{
+		TenantID: "organizations",
```
Comment (jhendrixMSFT) [inline sdk/azidentity/device_code_credential.go:40]: I assume this will be replaced with `organizationsTenantID` once your other PR is merged?

Comment (catalinaperalta) [reply sdk/azidentity/device_code_credential.go:40]: That is the plan. In the AuthCodeCredential PR

File: sdk/azidentity/device_code_credential.go (hunk 5 above)
Comment (catalinaperalta) [reply sdk/azidentity/device_code_credential.go:21]: I'll add this change now, but I will have to create a separate PR to do this for all of the other credentials as well.

Comment (catalinaperalta) [reply sdk/azidentity/device_code_credential.go]: Done

File: sdk/azidentity/device_code_credential.go [hunk 7]
```diff
@@ -15,27 +16,51 @@ const (
...
+
 // NewDeviceCodeCredential constructs a new DeviceCodeCredential used to authenticate against Azure Active Directory with a device code.
-// tenantID: The Azure Active Directory tenant (directory) ID of the service principal. If none is set then the default value ("organizations") will be used in place of the tenantID.
-// clientID: The client (application) ID of the service principal.
-// callback: The callback function used to send the login message back to the user
-// options: Options used to configure the management of the requests sent to Azure Active Directory.
-func NewDeviceCodeCredential(tenantID string, clientID string, callback func(string), options *TokenCredentialOptions) (*DeviceCodeCredential, error) {
-	c, err := newAADIdentityClient(options)
+// options: Options used to configure the management of the requests sent to Azure Active Directory, please see DeviceCodeCredentialOptions for a description of each field.
+func NewDeviceCodeCredential(options *DeviceCodeCredentialOptions) (*DeviceCodeCredential, error) {
```
Comment (jhendrixMSFT) [inline sdk/azidentity/device_code_credential.go:54]: Also add `var _ azcore.TokenCredential = (*DeviceCodeCredential)(nil)` someplace.

Comment (catalinaperalta) [reply sdk/azidentity/device_code_credential.go:54]: Done

File: sdk/azidentity/device_code_credential.go [hunk 8]
```diff
@@ -15,27 +16,51 @@ const (
...
 )
 
+// DeviceCodeCredentialOptions provide options that can configure DeviceCodeCredential instead of using the default values.
+type DeviceCodeCredentialOptions struct {
+	// Gets the Azure Active Directory tenant (directory) ID of the service principal
+	TenantID string
+	// Gets the client (application) ID of the service principal
+	ClientID string
+	// The callback function used to send the login message back to the user
+	UserPrompt func(string)
```
Comment (catalinaperalta) [inline sdk/azidentity/device_code_credential.go:26]: Is UserPrompt a clear enough name for the callback?

Comment (jhendrixMSFT) [reply sdk/azidentity/device_code_credential.go:26]: I was thinking about this.  Maybe just `Prompt`?  What do other SDKs call this?

Comment (catalinaperalta) [reply sdk/azidentity/device_code_credential.go:26]: Well there's a mix of names for the other langs:
NET - deviceCodeCallback
JAVA - challengeConsumer
JS/TS - userPromptCallback
PYTHON - prompt_callback

Comment (jhendrixMSFT) [reply sdk/azidentity/device_code_credential.go:26]: Leave as `UserPrompt` for now per discussion.

**Review by jhendrixMSFT — APPROVED** [review]
Replace the string with the const after resolving the merge conflicts.

**Pull request merged**